Tmp branch

Ekvo opened the pull request:
Realise final project. Scheduler - platinum. Docs iinside (link in README.md or ./docs/docs.go)

Commit:
Create:
dir init/ with .env file (TODO_PORT),
.golangci.yaml,
.gitignore: .idea/, lefthook.yml, env_to_test.txt,

Commit:
Create:
go.mod, go.sun,
main.go: http.Handler, http.ListenAndServ

Commit:
New dir:
internal/database:
1. transaction.go -> struct dbTX to accommodate *sql.Tx,
                      func (base *dbTX) Transaction(ctx context.Context, execute func(ctx context.Context) error) error,
2. database.go    -> container for dbTX from transaction.go,
                     InitDB() *sql.DB,
3. query.go       -> describes requests to database,
                     func (s Source) NewTables(ctx context.Context, tables ...string) error,
4. schema.go      -> describes table(s) in database,
                     const Schema,ctxTimeTableCreate

Commit:
New dir:
pkg/common:
common.go -> general utilities,
             CreatePathWithFile(partOfFilePath string) error

Commit:
.env (new - TODO_DBFILE), main.go db connect, test: app,db - ok

Commit:
New dir:
internal/server:
server.go -> wrapper to &gttp.Server,
main.go init server, test:app,db - ok

Commit:
New dir:
> internal/services:
> nextdate.go -> Realise algorithm for find task next date by day,year,weak and month - test: nextdate - ok FullNextDate = true,
pkg/common/common.go -> add func Abs(int)int

Commit:
main.go add handlerFunc for test 'NextDate' algorithm

Commit:
New dir: internal/model/ ->
> task.go:
> 1. TaskModel struct - describes table _scheduler_ in database look (internal/database/schema.go),
> 2. TaskCreate interface - rules for saving a task to storage,
> 3. SetDate(date string, algoNextDate func(time.Time, string, string) (string, error)) error - member TaskModel,
> 4. TaskComplete() (bool, error) - member TaskModel,
> 5. const: create -> TaskTitleLen, TaskCommentLen, TaskRepeatLen;  move -> DateFormat frpm internal/services/nextdate.go
internal/services/nextdate.go - minor changes

Commit:
internal/database:
query.go -> SaveOneTask(ctx context.Context, data any) (uint, error) member of Source, func WhenEmptyStringThenNULL(s string) *string,
database.go, transaction.go -> add comments, minor changes

Commit:
pkg/common/:
common.go -> type Message, String() string - Message member, var ErrCommonInvalidMedia

Commit:
New .go file: internal/services/taskdecode.go
rules for decode Task object,
1. TaskDecode struct,
2. NewTaskDecode() *TaskDecode,
3. Model() model.TaskModel - TaskDecode member,
4. Decode(r *http.Request, algoNextDate func(time.Time, string, string) (string, error)) error - TaskDecode member,
5. var: ErrServicesFiledEmpty, ErrServicesFiledLengthExceeded

Commit:
.env,lint, other - minor changes

Commit:
rebasing functions:
BeginningOfMonth(t time.Time) time.Time, ReduceTimeToDay(t time.Time) time.Time (old -  truncToDay(t time.Time) time.Time) from internal/services/nextdate.go to pkg/common/common.go

Commit:
common/common.go - SQlScan interface and minor changes

Commit:
model/task.go -> add TaskFind interface and minor changes

Commit:
New file: internal/services/taskencode.go
rules for serialize model.TaskModel to http.ResponseWriter

Commit:
New file: internal/services/taskproperty.go
TaskProperty struct - rules for create model.TaskModel array from database

Commit:
add to .gitignore

Commit:
database - minor changes

Commit:
database/query.go
model.TaskFind interface implementation - func (s Source) FindTaskList(ctx context.Context, data any) ([]model.TaskModel, error)

Commit:
New dir: internal/transport
1. transport.go - wrapper to *http.ServeMux and server.Srv - muxs and Run application,
2. handler.go - describes creating of routes group,

3. route.go - implementation of http.HandlerFunc of application

Commit:
main.go - code compression

Commit:
change tests/settings.go

Commit:
.github/workflows/tests.yml - set other test and go version 1.23

Commit:
Rebase (func SetData) from model/task.go -> services/taskdecode.go
task.go -> TaskModek member (func TaskComplete() (bool, error)) is deleted,
taskdecode.go -> TaskDecode now have member (executeDate(nextDate func(time.Time, string, string) (string, error)) (string, error)),
taskencode.go -> new TaskEncode strcut with member (Response() TaskResponse), changes in ( Response() []TaskResponse),
taskproperty.go -> minor changes

Commit:
common/common.go:
func EncodeJSON(ctx context.Context, w http.ResponseWriter, httpCode int, obj any) to
func EncodeJSON(w http.ResponseWriter, httpCode int, obj any)

Commit:
database/query.go:
new func:
1. FindOneTask(ctx context.Context, data any) (model.TaskModel, error) - realises model.TaskRead interface,
2.  NewDataTask(ctx context.Context, data any) error - realises model.TaskUpdate interface,
3. ExpirationTask(ctx context.Context, data any) error - realises model.TaskDelete interface

Commit:
internal/transport:
handler.go -> rename (TaskModelRoutes(db multiTask))->(taskRoutes(db multiTask)), add handleFuncs to taskRoutes, add into multiTask interfaces,
route.go -> realise: TaskDone(db multiTask) http.HandlerFunc, TaskRemove(db model.TaskDelete) http.HandlerFunc, TaskChange(db model.TaskUpdate) http.HandlerFunc, add var - ErrTransportInvalidParam,
transport.go -> minor changes

Commit:
New dir:  internal/services/deserializer, internal/services/serializer,
rebase -> internal/services/taskdecode.go -> internal/services/deserializer/taskdecode.go,
          internal/services/taskencode.go -> internal/services/serializer/taskencode.go

Commit:
New file: model/login.go, deserializer/logindecode.go ,serializer/loginencode.go:
login.go -> describes the Login object,
logindecode.go -> decode Login object,
loginencode.go -> encode Token after valid Login

Commit:
New dir: internal/services/autorization
new file autorization/autorization.go - midlweare functionб
taskdecode.go taskencode.go -> rename package

Commit:
common.go:
new func:
1.  HashData(line string) string - hashes a string,
2.  TokenGenerator(content string) (string, error) - create Bearer token with content inside,
3.  ReceiveValueFromToken[T any](token *jwt.Token, key string) (T, error) - check Bearer token and return value by key if exist,
4.  ReadCookie(r *http.Request, key string) (string, error),
5.  CleanCookie(w http.ResponseWriter, r *http.Request) - set MaxAge = -1

Commit:
route.go:
new func: func Login(w http.ResponseWriter, r *http.Request),
and minor changes

Commit:
handler.go:
add  mux.HandleFunc("POST /signin", Login),
and set  autorization.Autorization for every (/task) and (/tasks)

Commit:
nextdate.go: minor changes

Commit:
.env: add TODO_SECRET_KEY, TODO_PASSWORD

Commit:
go mod -> add github.com/golang-jwt/jwt/v5 v5.2.2, main.go ->func Init: common.SecretKey = os.Getenv(TODO_SECRET_KEY)

Commit:
settings.go:
all true, add Token -> test ./... is ok

Commit:
common.go: minor changes in DecodeJSON

Commit:
internal/database - minor changes

Commit:
internal/services - minor changes

Commit:
internal/server - change set default from int to string

Commit:
internal/transport:
handler.go -> change name taskRoutes to apiRoutes,
transport.go -> rename taskRoutes

Commit:
pkg/common -> common.go: new var ErrCommonEmptyBody, change in  DecodeJSON

Commit:
tests:
settings.go - Port to standart - 7540, DBFile to standart - ../scheduler.db, update Token, add func Init() - {find .env file}

Commit:
New dir: docs
docs.go - documentation

Commit:
.env file: new (TODO_VERSION)

Commit:
New file:
Dockerfile, compose.yaml,
.github/workflows/tests.yml add 21 line - uses: golangci/golangci-lint-action@v3

Commit:
README.md - major changes, .gitignore (remove - olddata)

Commit:
New dir: storage

Commit:
init/.env -> add var TODO_TEST_DBFILE, internal/database/database.go -> change InitDB major, main.go change call InitDB

Commit:
transport/route.go -> major change TaskChange(db model.TaskUpdate) http.HandlerFunc, minor change TaskNew(db model.TaskCreate) http.HandlerFunc

Commit:
C:/Program Files/Git/common/common.go -> major change func (m Message) String() string

Commit:
services/nextdate.go minor change monthAndDay(repeat string) ([]any, error), services/deserializer/taskdecode.go comments

Commit:
New tests files:
database/database_test.go,
services/autorization/autorization_test.go,
deserializer/deserializer_test.go,
serializer/serializer_test.go,
services/services_test.go,
transport/transport_test.go,
common/common_test.go

Commit:
New files - coverage after test

Commit:
README.md

ivannizh commented:
При открытии страницы сразу 403 ошибка
![image](https://github.com/user-attachments/assets/14f1dbf4-aaa9-4776-8a90-e9cd26a9dd21)
 Ошибка в тестах
![image](https://github.com/user-attachments/assets/6113e81e-59e3-45fc-825f-10b836a85cee)
У меня часовой пояс GMT+7 и время 00:14. Тесты могут верно брать мой новый день, а код читает по UTC

Ekvo commented:
http://127.0.0.1:8000/login.html
qwert12345

сб, 12 апр. 2025 г. в 20:15, Ivan Nizhnichenko ***@***.***>:

> При открытии страницы сразу 403 ошибка
> image.png (view on web)
> <https://github.com/user-attachments/assets/14f1dbf4-aaa9-4776-8a90-e9cd26a9dd21>
> Ошибка в тестах
> image.png (view on web)
> <https://github.com/user-attachments/assets/6113e81e-59e3-45fc-825f-10b836a85cee>
> У меня часовой пояс GMT+7 и время 00:14. Тесты могут верно брать мой новый
> день, а код читает по UTC
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ekvo/yandex-practicum-go-final-project/pull/1#issuecomment-2798912259>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BK2QKMZM536VPPFU3SNFAAT2ZFCZZAVCNFSM6AAAAAB3AFV7AGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOJYHEYTEMRVHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> *ivannizh* left a comment (Ekvo/yandex-practicum-go-final-project#1)
> <https://github.com/Ekvo/yandex-practicum-go-final-project/pull/1#issuecomment-2798912259>
>
> При открытии страницы сразу 403 ошибка
> image.png (view on web)
> <https://github.com/user-attachments/assets/14f1dbf4-aaa9-4776-8a90-e9cd26a9dd21>
> Ошибка в тестах
> image.png (view on web)
> <https://github.com/user-attachments/assets/6113e81e-59e3-45fc-825f-10b836a85cee>
> У меня часовой пояс GMT+7 и время 00:14. Тесты могут верно брать мой новый
> день, а код читает по UTC
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ekvo/yandex-practicum-go-final-project/pull/1#issuecomment-2798912259>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BK2QKMZM536VPPFU3SNFAAT2ZFCZZAVCNFSM6AAAAAB3AFV7AGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOJYHEYTEMRVHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Ekvo commented:
да cookie действительны 5-6 дней
в общем документация
ссылка а README.md


сб, 12 апр. 2025 г. в 20:19, Александр ***@***.***>:

> http://127.0.0.1:8000/login.html
> qwert12345
>
> сб, 12 апр. 2025 г. в 20:15, Ivan Nizhnichenko ***@***.***>:
>
>> При открытии страницы сразу 403 ошибка
>> image.png (view on web)
>> <https://github.com/user-attachments/assets/14f1dbf4-aaa9-4776-8a90-e9cd26a9dd21>
>> Ошибка в тестах
>> image.png (view on web)
>> <https://github.com/user-attachments/assets/6113e81e-59e3-45fc-825f-10b836a85cee>
>> У меня часовой пояс GMT+7 и время 00:14. Тесты могут верно брать мой
>> новый день, а код читает по UTC
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Ekvo/yandex-practicum-go-final-project/pull/1#issuecomment-2798912259>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/BK2QKMZM536VPPFU3SNFAAT2ZFCZZAVCNFSM6AAAAAB3AFV7AGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOJYHEYTEMRVHE>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>> *ivannizh* left a comment (Ekvo/yandex-practicum-go-final-project#1)
>> <https://github.com/Ekvo/yandex-practicum-go-final-project/pull/1#issuecomment-2798912259>
>>
>> При открытии страницы сразу 403 ошибка
>> image.png (view on web)
>> <https://github.com/user-attachments/assets/14f1dbf4-aaa9-4776-8a90-e9cd26a9dd21>
>> Ошибка в тестах
>> image.png (view on web)
>> <https://github.com/user-attachments/assets/6113e81e-59e3-45fc-825f-10b836a85cee>
>> У меня часовой пояс GMT+7 и время 00:14. Тесты могут верно брать мой
>> новый день, а код читает по UTC
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Ekvo/yandex-practicum-go-final-project/pull/1#issuecomment-2798912259>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/BK2QKMZM536VPPFU3SNFAAT2ZFCZZAVCNFSM6AAAAAB3AFV7AGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOJYHEYTEMRVHE>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

Commit:
Major: all time delte UTC() -> UTC only with time.After and time.Before

Ekvo commented:
сб, 12 апр. 2025 г. в 20:22, Александр ***@***.***>:

> да cookie действительны 5-6 дней
> в общем документация
> ссылка а README.md
>
>
> сб, 12 апр. 2025 г. в 20:19, Александр ***@***.***>:
>
>> http://127.0.0.1:8000/login.html
>> qwert12345
>>
>> сб, 12 апр. 2025 г. в 20:15, Ivan Nizhnichenko ***@***.***
>> >:
>>
>>> При открытии страницы сразу 403 ошибка
>>> image.png (view on web)
>>> <https://github.com/user-attachments/assets/14f1dbf4-aaa9-4776-8a90-e9cd26a9dd21>
>>> Ошибка в тестах
>>> image.png (view on web)
>>> <https://github.com/user-attachments/assets/6113e81e-59e3-45fc-825f-10b836a85cee>
>>> У меня часовой пояс GMT+7 и время 00:14. Тесты могут верно брать мой
>>> новый день, а код читает по UTC
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/Ekvo/yandex-practicum-go-final-project/pull/1#issuecomment-2798912259>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/BK2QKMZM536VPPFU3SNFAAT2ZFCZZAVCNFSM6AAAAAB3AFV7AGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOJYHEYTEMRVHE>
>>> .
>>> You are receiving this because you authored the thread.Message ID:
>>> ***@***.***>
>>> *ivannizh* left a comment (Ekvo/yandex-practicum-go-final-project#1)
>>> <https://github.com/Ekvo/yandex-practicum-go-final-project/pull/1#issuecomment-2798912259>
>>>
>>> При открытии страницы сразу 403 ошибка
>>> image.png (view on web)
>>> <https://github.com/user-attachments/assets/14f1dbf4-aaa9-4776-8a90-e9cd26a9dd21>
>>> Ошибка в тестах
>>> image.png (view on web)
>>> <https://github.com/user-attachments/assets/6113e81e-59e3-45fc-825f-10b836a85cee>
>>> У меня часовой пояс GMT+7 и время 00:14. Тесты могут верно брать мой
>>> новый день, а код читает по UTC
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/Ekvo/yandex-practicum-go-final-project/pull/1#issuecomment-2798912259>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/BK2QKMZM536VPPFU3SNFAAT2ZFCZZAVCNFSM6AAAAAB3AFV7AGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOJYHEYTEMRVHE>
>>> .
>>> You are receiving this because you authored the thread.Message ID:
>>> ***@***.***>
>>>
>>

Ekvo commented:
Открыть страницу без авторизации невозможною. Это часть задания там midlweare.
Документация, UTC переделал

Commit:
transport/route.go -> Login(w http.ResponseWriter, r *http.Request) -> setcookie

Ekvo commented:
Убедительная просба вводить пароль перед проверкой.
http://127.0.0.1:8000/login.html 
password:qwert12345

вы логинетесь, создаете cookei и дальше проходите через авторизацию

internal/services/autorization/autorization.go

Ekvo commented:
UTC проверил,  я бы не догадался. С cookie вышла недоразумение, они были в какой то момент мной закомиченны
 - UTC убрал, ошибка пропала (проверял - была)
-  cookie раскоммитил.

вход:
http://127.0.0.1:8000/login.html
password:
```
qwert12345
```

удалял cookie в браузере, и сетал через  Login. Все работает.

спасибо за UTC)

ivannizh commented:
Давай по ТЗ, я вхожу на главный сайт, а он уже делает редирект
![image](https://github.com/user-attachments/assets/2dc432ab-e6da-4b1d-ad4a-943db89eb99c)

Commit:
autorization/autorization.go -> major changes: AuthZ - add redirect, README.md change Test bash code

Commit:
C:/Program Files/Git/autorization_test.go - pass

Ekvo commented:
Поправил.

password:
```
qwert12345
```

Commit:
C:/Program Files/Git/autorization/autorization.go -> remove redirect

Commit:
docs/docs.go -> finish start line IMPORTANT: NOT TIME READ DOCS,README.md - add curl into Test

Commit:
model/task.go -> Major changes int UpdateDate member of TaskModel

Commit:
database/query.go:
SaveOneTask(ctx context.Context, data any) (uint, error) now - return err,
FindOneTask(ctx context.Context, data any) (model.TaskModel, error) - if err != nil && errors.Is(err, sql.ErrNoRows),
NewDataTask(ctx context.Context, data any) error - if err != nil && errors.Is(err, sql.ErrNoRows),
ExpirationTask(ctx context.Context, data any) error - if err != nil && errors.Is(err, sql.ErrNoRows),
var ErrDataBaseAlreadyExist  - deleted

ivannizh commented:
Нет сортировки
![image](https://github.com/user-attachments/assets/926112fb-c036-48ef-b283-544453b4873d)

ivannizh commented on Dockerfile:
```diff
@@ -0,0 +1,43 @@
+FROM golang:1.23.0 AS builder
```
команды можно объединять, так удобнее читать
![image](https://github.com/user-attachments/assets/d1088162-30ac-4540-81a3-05cc33f3a5cd)
![image](https://github.com/user-attachments/assets/d3cb7199-e584-49dd-8375-ef248114dbec)

Ekvo commented:
Пожалуйста скажите мне, какие команды вы вводите, я не понимаю, что мне исправлять, я в замешательстве.

Ekvo commented:
![изображение](https://github.com/user-attachments/assets/32161605-e270-47fc-b86b-420907235bc0)

ivannizh commented on README.md:
Отличный ридми!

ivannizh commented on internal/database/coverage:
Обычно такие файлы не кладут в репу

Ekvo replied:
запомнил

ivannizh commented on main.go:
По твоей структуре файл должен тогда лежать в `cmd/tracker/main.go`

Ekvo replied:
да я думал об этом

ivannizh commented on main.go:
```diff
@@ -0,0 +1,39 @@
+	}
+}
+
+func main() {
```
Отличный лаконичный main

ivannizh commented on internal/database/database.go:
```diff
@@ -0,0 +1,56 @@
+	install := false
+	if _, err := os.Stat(pathDB); err != nil {
+		if err := common.CreatePathWithFile(pathDB); err != nil {
+			return nil, fmt.Errorf("database: file.db create error - %v", err)
```
При оборачивании нужно использовать `%w` (wrap), а просто при выводе в консоль `%v`
```suggestion
			return nil, fmt.Errorf("database: file.db create error - %w", err)
```

ivannizh replied:
Нужно везде поправить

ivannizh commented on internal/database/query.go:
```diff
@@ -0,0 +1,159 @@
+		}
+		return err
+	}
+	return s.store.Transaction(ctx, updateTask)
```
Не очень понимаю зачем транзакция для одной операции

ivannizh replied:
Вообще в этой работе транзакции не нужны.
Можно их сделать, но зачем

ivannizh commented on internal/database/query.go:
```diff
@@ -0,0 +1,159 @@
+
+	query.WriteString("SELECT * FROM scheduler")
+	if property.IsWord() {
+		query.WriteString("\nWHERE title LIKE $1 OR comment LIKE $1\nORDER BY date ASC")
```
Хорошо сделано, но зачем переносы на новую строку, БД так поймет нас

Ekvo commented:
> Нет сортировки ![image](https://private-user-images.githubusercontent.com/9384832/433142166-926112fb-c036-48ef-b283-544453b4873d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ1NTYzODUsIm5iZiI6MTc0NDU1NjA4NSwicGF0aCI6Ii85Mzg0ODMyLzQzMzE0MjE2Ni05MjYxMTJmYi1jMDM2LTQ4ZWYtYjI4My01NDQ0NTNiNDg3M2QucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MTNUMTQ1NDQ1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjIyYWZmMzM4MTI1OGRiMWYzYWYwNDJkZGZmMWEwNjIzM2U5Y2JkYWI4ZjdhMDNmMmNmZDZmZGY2ODc4MGZjMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.C2MtzYULFtDKkGj9P50UcvFvPTYo3zhqk5bBZZk48tk)

нашел ошибку
func (s Source) FindTaskList(ctx context.Context, data any) ([]model.TaskModel, error)
пакет database/query.go

Commit:
database/query.go -> minor change FindTaskList(ctx context.Context, data any) ([]model.TaskModel,error)

ivannizh commented on internal/services/autorization/autorization.go:
```diff
@@ -0,0 +1,61 @@
+//     write to log line content
+//
+//  3. call next(w,r)
+func AuthZ(next http.HandlerFunc) http.HandlerFunc {
```
Знаю, что в ТЗ так написано, но большой иф плохо, можно просто так сделать и код понятнее читать
![image](https://github.com/user-attachments/assets/b2a07d64-0758-4b52-87f0-160d4fb3bc6a)

Ekvo commented:
Исправил


вс, 13 апр. 2025 г. в 17:53, Ivan Nizhnichenko ***@***.***>:

> Нет сортировки
> image.png (view on web)
> <https://github.com/user-attachments/assets/926112fb-c036-48ef-b283-544453b4873d>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ekvo/yandex-practicum-go-final-project/pull/1#issuecomment-2799984157>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BK2QKM6U53SMQL3ZVNG4VN32ZJ3AFAVCNFSM6AAAAAB3AFV7AGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOJZHE4DIMJVG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> *ivannizh* left a comment (Ekvo/yandex-practicum-go-final-project#1)
> <https://github.com/Ekvo/yandex-practicum-go-final-project/pull/1#issuecomment-2799984157>
>
> Нет сортировки
> image.png (view on web)
> <https://github.com/user-attachments/assets/926112fb-c036-48ef-b283-544453b4873d>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ekvo/yandex-practicum-go-final-project/pull/1#issuecomment-2799984157>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BK2QKM6U53SMQL3ZVNG4VN32ZJ3AFAVCNFSM6AAAAAB3AFV7AGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOJZHE4DIMJVG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

ivannizh commented on internal/services/autorization/autorization_test.go:
```diff
@@ -0,0 +1,75 @@
+	}
+}
+
+func TestAuthZ(t *testing.T) {
```
У тебя нормально протестировать не получится мидлаварю, потому что она не чистая (читай про "Чистая функция") из-за пароля, это внешняя зависимость внутри функции
![image](https://github.com/user-attachments/assets/095e5d5c-777c-4cf1-8142-3e23b7d2a203)
Вот так можно сделать и можно уже тестировать
![image](https://github.com/user-attachments/assets/74bb12a1-0a85-4a76-b09e-6939a609e76a)

ivannizh commented on internal/services/serializer/serializer_test.go:
```diff
@@ -0,0 +1,82 @@
+	if err := godotenv.Load("../../../init/.env"); err != nil {
+		log.Printf("autorization_test: no .env file - %v", err)
+	}
+	common.SecretKey = os.Getenv("TODO_SECRET_KEY")
```
Не очень понял зачем это.

Вообще в тестах это выглядит не норм, у тебя есть файлик с энвом и энв, от которого зависят тесты.

Лучше сделать пакет config, который по умолчанию инициализируется из эва, а для тестов есть функция, которая собирает тебе нужный конфиг, так у тебя тесты будут детерминированы.

ivannizh commented on internal/transport/route.go:
```diff
@@ -0,0 +1,186 @@
+	password := os.Getenv("TODO_PASSWORD")
+	login := deserialize.Model()
+	if !login.ValidPassword(password) {
+		common.EncodeJSON(w, http.StatusForbidden, common.Message{"error": model.ErrModelsLoginInvalidPassword.Error()})
```
401

ivannizh commented on internal/transport/route.go:
```diff
@@ -0,0 +1,186 @@
+func Login(w http.ResponseWriter, r *http.Request) {
+	deserialize := deserializer.NewLoginDecode()
+	if err := deserialize.Decode(r); err != nil {
+		common.EncodeJSON(w, http.StatusUnprocessableEntity, common.Message{"error": err.Error()})
```
А у нас тут в теле пришла каша и мы не поняли что это такое, так что тут скорее 400

![image](https://github.com/user-attachments/assets/71b65dff-d3ea-43ab-b828-f3855d7f28e1)

ivannizh commented on internal/transport/route.go:
```diff
@@ -0,0 +1,186 @@
+		}
+		taskID, err := db.SaveOneTask(r.Context(), deserialize.Model())
+		if err != nil {
+			common.EncodeJSON(w, http.StatusConflict, common.Message{"error": err.Error()})
```
Тут скорее внутренняя ошибка, например, отвалилась БД. Если ты имел в виду ошибки, что такой айди уже есть, то нужно отдельно это обработать, но у нас он автоматический. 
Так что тут 500
![image](https://github.com/user-attachments/assets/95bbc17f-28f7-4551-9829-f79acd57f3c7)

ivannizh commented on internal/transport/route.go:
```diff
@@ -0,0 +1,186 @@
+	}
+}
+
+func TaskRetrive(db model.TaskRead) http.HandlerFunc {
```
```suggestion
func TaskRetrieve(db model.TaskRead) http.HandlerFunc {
```

ivannizh commented on internal/transport/route.go:
```diff
@@ -0,0 +1,186 @@
+		}
+		task, err := db.FindOneTask(r.Context(), uint(id))
+		if err != nil {
+			common.EncodeJSON(w, http.StatusNotFound, common.Message{"error": err.Error()})
```
Так же нужно разделять, что прислала БД, нет задачи или проблема с запросом

ivannizh replied:
далее везде так же

Ekvo replied:
это важно, спасибо.

ivannizh commented on internal/transport/route.go:
```diff
@@ -0,0 +1,186 @@
+	common.EncodeJSON(w, http.StatusOK, tokenResponse)
+}
+
+func TaskNew(db model.TaskCreate) http.HandlerFunc {
```
Твоему коду не хватает Usecase слоя

https://habr.com/ru/articles/887102/

Ekvo replied:
До сего не знал об этом, буду изучат.
Благодарю.

ivannizh commented on internal/transport/transport_test.go:
```diff
@@ -0,0 +1,339 @@
+	*s.id++
+}
+
+func (s mockStore) SaveOneTask(_ context.Context, data any) (uint, error) {
```
Кажется, что ты это писал руками, потом можно освоить пару либ, которые генерируют тебе моки, удобные для использовании в тестах.

Ekvo replied:
Да, верно, писал руками))

ivannizh commented on pkg/common/common.go:
```diff
@@ -0,0 +1,210 @@
+//
+// sort keys - result was predictable
+func (m Message) String() string {
+	buff := &bytes.Buffer{}
```
мудрено с буфером
![image](https://github.com/user-attachments/assets/1134fd21-7efb-4839-9f11-2898bfdaadf8)

Можно даже так
![image](https://github.com/user-attachments/assets/57425e9c-d6c5-428a-ba0c-3b55b0da0ad8)

Ekvo replied:
да, так короче - лучше

ivannizh reviewed (approved):
Александр, привет!

Отличная работа, всё запускается, тесты проходят.
Достойная работа, видно, что сам много где разбирался.
Код хорошо раскидан по файлам, отдельный лайк за свои тесты!
Работу можно еще улучшать и улучшать, постарался оставить побольше комментариев.

Поздравляю с окончанием курса и желаю успехов в дальнейшем освоении Go!

Commit:
Start sheduler v2.0.0:
main.go -> cmd/app/main.go,
init/.env -> new variables,
internal/services/nextdate.go -> internal/lib/nextdate/nextdate.go,
internal/services/taskproperty.go -> internal/services/entity/taskproperty.go,
delete: all coverage,
storage/scheduler.db,
internal/services/autorization/autorization_test.go, internal/services/autorization/autorization.go,

modified:
internal/services/deserializer/taskdecode.go ,
internal/services/deserializer/logindecode.go,
internal/services/deserializer/deserializer_test.go

Commit:
New dir: internal/lib:
jwtsign/jwtsign.go -> contain secretKey for jwt.Token and func service this key,
jwtsign_test.go,
 pkg/common/common.go -> delete all everything related with jwt.Token, add MessageError, change in common_test.go

Commit:
internal/lib: add tests

Commit:
internal/services/entity: add tests

Commit:
internal/services/deserializer: minor changes

Commit:
internal/services/serializer:
(t TokenEncode) Response() - now return (*TokenResponse, error),
taskencode.go:
add: TaskIDResponse struct,  TaskIDEncode struct, (tIDe TaskIDEncode) Response() *TaskIDResponse -> create ID of task for Response,
now we return only ptr of objexts for Response

Commit:
New file internal/services/services.go: describes all biz logic of application, contain only interfaces

Commit:
New dir: internal/services/usecase:
usecase/authcase.go - service of autorization,
usecase/logincase.go - service of user,
usecase/taskcase.go  - service of task,
usecase/usecase_test.go - test for package usecase

Commit:
New dir: internal/config:
config.go - get data from .env file or ENV variables for start application,

options.go - describes parse file options for config

Commit:
internal/database:
database.go - major change in func InitDB now take cfg *config.Config,
query.go - change in FindTaskList

Commit:
New dir: internal/datauser:
datauser.go -> cache store for user password set with help cf *config.Config

Commit:
server/server.go now create with help cfg *config.Config

Commit:
New file: transport/midlweare.go - describes midlweare function for routes (add func AuthZ and interface rulesForAuthZ)

Commit:
internal/transport:
handler.go -> delete multiTask interface, change old autorization.AuthZ for midlweare.AuthZ,
transport.go -> add shedulerCase interface - contain all interface from services/services.go, Run - Start,
route.go -> MAJOR changes - login,task handlers - now have biz logic of application,
transport_test.go -> add tests

Commit:
internal/model:
login.go -> add LoginRead interface - describes rule for read data user from store,
task.go -> member of TaskModel UpdateDate - is deleted

Commit:
New dir: internal/app:
app.go -> heart of application contain all biz logic,
run.go -> Run function - start and closed application

Commit:
go.mod go.sum

Commit:
.gitignore

Commit:
cmd/app/main.go: delete func init(), now ccreate config and call Run from internal/app/run.go

Commit:
init/.env:
delete TODO_TEST_DBFILE, TODO_TEST_SECRET_KEY,
change TODO_WEB -> PATH_DIR_WEB,
update TODO_VERSION

Commit:
Update Dokcerfile

Commit:
complete docs/docs.go and README.md

Commit:
tests: settings.go -> update Token (21.04.25, life - 7 days)

Commit:
config/config.go: (cfg *Config) validPassword(msgErr common.Message) - commented out

Commit:
usecase/usecase_test.go - minor change

Commit:
.env now correct

Commit:
README.md - minor change

Commit:
workflows/tests.yml -> set file for run
